### ISSUE TITLE
Docker no starting on WORKER in swarm mode

### ISSUE THREAD

**Issue by rvernica** [docker/docker#24078]
**Output of `docker version` on MANAGER:**

```
> docker version
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info` on MANAGER:**

```
> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.12.0-rc2
Storage Driver: devicemapper
 Pool Name: centos-dockerpool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 2.541 GB
 Data Space Total: 42.95 GB
 Data Space Available: 40.41 GB
 Metadata Space Used: 3.67 MB
 Metadata Space Total: 41.94 MB
 Metadata Space Available: 38.27 MB
 Thin Pool Minimum Free Space: 4.295 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null overlay bridge host
Swarm: active
 NodeID: bcunjya7hjuhdlyn3odnx3e1q
 IsManager: Yes
 Managers: 1
 Nodes: 1
 CACertHash: sha256:03af27975890f7c5b9b64abb2723fec74dd565ea97a4e3559f7695c282e7a71a
Runtimes: default
Default Runtime: default
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 23.37 GiB
Name: blu-03
ID: CBXH:UD64:UUH7:DK4P:D2AC:Y627:3EGT:OTYB:YQTC:QH5C:TZKV:Q7X7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://...
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Three physical machines (MANAGER and two WORKERS) running CentOS 7.2 with `firewalld` disabled. They have the same Docker version and setup.
```
> cat /etc/centos-release
CentOS Linux release 7.2.1511 (Core) 
> uname -a
Linux blu-03 3.10.0-327.22.2.el7.x86_64 #1 SMP Thu Jun 23 17:05:11 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
> sudo service firewalld status
Redirecting to /bin/systemctl status  firewalld.service
● firewalld.service - firewalld - dynamic firewall daemon
   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; disabled; vendor preset: enabled)
   Active: inactive (dead)
```

**Steps to reproduce the issue:**
1. On one of the machines (MANAGER), do `docker swarm init`
2. On the other two machines (WORKER), do `docker swarm join <MANAGER-IP>`
3. Restart all three machines

**Describe the results you received:**
Docker no longer starts on the WORKER machines. (Docker starts fine on MANAGER machine.)
```
> sudo service docker status -l
Redirecting to /bin/systemctl status  -l docker.service
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/docker.service.d
           └─http-proxy.conf, storage.conf
   Active: failed (Result: exit-code) since Tue 2016-06-28 16:08:42 PDT; 13min ago
     Docs: https://docs.docker.com
  Process: 980 ExecStart=/usr/bin/dockerd -H fd:// --storage-driver devicemapper --storage-opt dm.thinpooldev=/dev/mapper/centos-dockerpool (code=exited, status=1/FAILURE)
 Main PID: 980 (code=exited, status=1/FAILURE)

Jun 28 16:08:32 blu-04 dockerd[980]: time="2016-06-28T16:08:32.336448954-07:00" level=warning msg="Running modprobe bridge br_netfilter failed with message: modprobe: WARNING: Module br_netfilter not found.\ninsmod /lib/modules/3.10.0-327.22.2.el7.x86_64/kernel/net/llc/llc.ko \ninsmod /lib/modules/3.10.0-327.22.2.el7.x86_64/kernel/net/802/stp.ko \ninsmod /lib/modules/3.10.0-327.22.2.el7.x86_64/kernel/net/bridge/bridge.ko \n, error: exit status 1"
Jun 28 16:08:32 blu-04 dockerd[980]: time="2016-06-28T16:08:32.364210543-07:00" level=info msg="Firewalld running: false"
Jun 28 16:08:32 blu-04 dockerd[980]: time="2016-06-28T16:08:32.603355256-07:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Jun 28 16:08:32 blu-04 dockerd[980]: time="2016-06-28T16:08:32.698269276-07:00" level=info msg="Loading containers: done."
Jun 28 16:08:42 blu-04 dockerd[980]: time="2016-06-28T16:08:42.069860930-07:00" level=error msg="cluster exited with error: can't create raft node: rpc error: code = 2 desc = grpc: the client connection is closing"
Jun 28 16:08:42 blu-04 dockerd[980]: time="2016-06-28T16:08:42.069987542-07:00" level=fatal msg="Error creating cluster component: swarm component could not be started"
Jun 28 16:08:42 blu-04 systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Jun 28 16:08:42 blu-04 systemd[1]: Failed to start Docker Application Container Engine.
Jun 28 16:08:42 blu-04 systemd[1]: Unit docker.service entered failed state.
Jun 28 16:08:42 blu-04 systemd[1]: docker.service failed.
```

**Describe the results you expected:**
WORKER machines can join the swarm and Docker starts on WORKER machines.

**Additional information you deem important (e.g. issue happens only occasionally):**
Initially `firewalld` was running. Once I stopped `firewalld`, one of the WORKERS was able to join the swarm, `docker node ls` on MANAGER listed the WORKER. This happened once and I am not able to reproduce it. I disabled `firewalld` and rebooted the machines before reporting this.

**Comment by nathanleclaire** [comment]
Hi @rvernica, thanks for the issue.

> Initially firewalld was running. Once I stopped firewalld, one of the WORKERS was able to join the swarm, docker node ls on MANAGER listed the WORKER. This happened once and I am not able to reproduce it. I disabled firewalld and rebooted the machines before reporting this.

I think it sounds like network issue.  If it is not reproducible maybe @tiborvass would like to close.

FWIW, in order for Swarm to run effectively you _generally_ need:

- Inbound `2377` access allowed from worker to manager nodes
- Inbound `4789` (VXLan) from all nodes to all other nodes for overlay networking to work

**Comment by rvernica** [comment]
The three machines I am using are connected to the same switch and I am connect to them using SSH. The SSH connection is stable and I do not believe there are any transient network issues.

The issue I am reporting is that Docker service not starting and this is reproducible every time. What I could not reproduce is getting them connected to the swarm.

I suspected there might be an issue with the inbound ports you listed, but since `firewalld` is off, there should not be anything else blocking the ports. The OS on the nodes is a minimal CentOS installed a few days ago.

Could you confirm that rc2 in swarm mode works well in CentOS with physical machines?

**Comment by nathanleclaire** [comment]
> The issue I am reporting is that Docker service not starting and this is reproducible every time. What I could not reproduce is getting them connected to the swarm.

I see.  In the "expected results" above it says:

> WORKER machines can join the swarm and Docker starts on WORKER machines.

So I wasn't aware that you had an issue with service start.

Can you provide some information about the service start issue?  And preferably the Docker daemon logs from the instances as well?

**Comment by rvernica** [comment]
Please see *Describe the results you received* above for the service start logs. For Docker daemon logs see [grep docker /var/log/messages](https://github.com/docker/docker/files/340049/docker-log.txt)

**Comment by nathanleclaire** [comment]
Thanks @rvernica .  We still need detailed steps on what issue you are running into exactly with service start, preferably with a minimally reproducible example.  What is the issue with service start?

**Comment by rvernica** [comment]
This is from the WORKER node:
```
> sudo service docker start
Redirecting to /bin/systemctl start  docker.service
Job for docker.service failed because the control process exited with error code. See "systemctl status docker.service" and "journalctl -xe" for details.
```
```
> sudo systemctl status docker.service -l
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/docker.service.d
           └─http-proxy.conf, storage.conf
   Active: failed (Result: exit-code) since Wed 2016-06-29 16:13:59 PDT; 1min 42s ago
     Docs: https://docs.docker.com
  Process: 3262 ExecStart=/usr/bin/dockerd -H fd:// --storage-driver devicemapper --storage-opt dm.thinpooldev=/dev/mapper/centos-dockerpool (code=exited, status=1/FAILURE)
 Main PID: 3262 (code=exited, status=1/FAILURE)

Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.425755320-07:00" level=info msg="Loading containers: start."
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.455316188-07:00" level=info msg="Firewalld running: false"
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.655138551-07:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.734657515-07:00" level=info msg="Loading containers: done."
Jun 29 16:13:58 blu-04 dockerd[3262]: time="2016-06-29T16:13:58.981324114-07:00" level=error msg="cluster exited with error: can't create raft node: rpc error: code = 2 desc = grpc: the client connection is closing"
Jun 29 16:13:58 blu-04 dockerd[3262]: time="2016-06-29T16:13:58.981437693-07:00" level=fatal msg="Error creating cluster component: swarm component could not be started"
Jun 29 16:13:59 blu-04 systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Jun 29 16:13:59 blu-04 systemd[1]: Failed to start Docker Application Container Engine.
Jun 29 16:13:59 blu-04 systemd[1]: Unit docker.service entered failed state.
Jun 29 16:13:59 blu-04 systemd[1]: docker.service failed.
```
```
> sudo journalctl -xe
-- Unit docker.socket has begun starting up.
Jun 29 16:13:47 blu-04 systemd[1]: Listening on Docker Socket for the API.
-- Subject: Unit docker.socket has finished start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit docker.socket has finished starting up.
-- 
-- The start-up result is done.
Jun 29 16:13:47 blu-04 systemd[1]: Starting Docker Application Container Engine...
-- Subject: Unit docker.service has begun start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit docker.service has begun starting up.
Jun 29 16:13:47 blu-04 dockerd[3262]: time="2016-06-29T16:13:47.306873197-07:00" level=info msg="New containerd process, pid: 3269"
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.382982700-07:00" level=warning msg="devmapper: Base device already exists and has filesystem xfs on it. User specified filesystem  will be ignored."
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.424433070-07:00" level=info msg="Graph migration to content-addressability took 0.00 seconds"
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.425236656-07:00" level=warning msg="mountpoint for pids not found"
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.425755320-07:00" level=info msg="Loading containers: start."
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.455316188-07:00" level=info msg="Firewalld running: false"
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.655138551-07:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Jun 29 16:13:48 blu-04 dockerd[3262]: time="2016-06-29T16:13:48.734657515-07:00" level=info msg="Loading containers: done."
Jun 29 16:13:58 blu-04 dockerd[3262]: time="2016-06-29T16:13:58.981324114-07:00" level=error msg="cluster exited with error: can't create raft node: rpc error: code = 2 desc = grpc: the client connection is closing"
Jun 29 16:13:58 blu-04 dockerd[3262]: time="2016-06-29T16:13:58.981437693-07:00" level=fatal msg="Error creating cluster component: swarm component could not be started"
Jun 29 16:13:59 blu-04 systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Jun 29 16:13:59 blu-04 systemd[1]: Failed to start Docker Application Container Engine.
-- Subject: Unit docker.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit docker.service has failed.
-- 
-- The result is failed.
```
The steps I did are listed in *Steps to reproduce the issue* in the issue description. Since Docker does not start for me, I can't follow the steps right now. I am kind-of stuck.

**Comment by thaJeztah** [comment]
pin @nathanleclaire

**Comment by nathanleclaire** [comment]
Hi @rvernica .  Thanks for the followup. I had misunderstood "systemd service" as "docker service" (as in the `docker service` command).

I would bet that this is a network issue of some kind.  Make 100% sure that you have connectivity to the inbound Swarm manager node and port from the workers.  Also, critically, make 100% certain that you are setting `--listen-addr` to the correct value for the manager node.  By default it aims to listen on all interfaces (`0.0.0.0`) but may fall apart in advertising the correct node IPs since it tries to divine them from the existing `eth0`, etc. interfaces if they are not specified directly (you'll notice that `0.0.0.0` is _never_ the right IP to advertise, so Swarm mode in 1.12.0-rc3 tried to figure it out for you -- I'm not sure the status quo of this fix on master, I think it errors out if there is ambiguity).  Your error looks like what I've gotten when I've had this issue.

**Comment by tiborvass** [comment]
@nathanleclaire #24237  ?

**Comment by nathanleclaire** [comment]
@tiborvass Bingo.

**Comment by rvernica** [comment]
@nathanleclaire I gave it another try. I upgraded to `rc4`:
```
03:~> docker version
Client:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        
 OS/Arch:      linux/amd64
```
I specified IP and port when initiating the swarm:
```
03:~> docker swarm init --listen-addr ...4.63:2377
No --secret provided. Generated random secret:
        ...

Swarm initialized: current node (5yjryq2sup5qbpmptegbg78o6) is now a manager.

To add a worker to this swarm, run the following command:
        docker swarm join --secret ... \
        --ca-hash sha256:... \
        ....4.63:2377
```
I successfully connected one worker:
```
04:~> docker swarm join --secret ... \
>         --ca-hash ... \
>         ...4.63:2377
This node joined a Swarm as a worker.
```
I can't connect a second worker:
```
05:~> docker swarm join --secret ... \
>         --ca-hash ... \
>         ....4.63:2377
Error response from daemon: Timeout was reached before node was joined. Attempt to join the cluster will continue in the background. Use "docker info" command to see the current Swarm status of your node.
```
`docker info` on the second worker reports:
```
05:~> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.0-rc4
Storage Driver: devicemapper
 Pool Name: centos-dockerpool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 20.25 MB
 Data Space Total: 42.91 GB
 Data Space Available: 42.89 GB
 Metadata Space Used: 217.1 kB
 Metadata Space Total: 41.94 MB
 Metadata Space Available: 41.73 MB
 Thin Pool Minimum Free Space: 4.291 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: pending
 NodeID: 
 IsManager: No
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 5.65 GiB
Name: ...05
ID: EM7E:PHL2:VOHH:TCP5:GID6:QKTH:WPHH:NDDV:HH5H:6LQ2:I6TC:ZU4C
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: ...
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```
Second worker can access the master on the `2377` port:
```
05:~> ssh -v -p 2377 ...4.63
OpenSSH_6.6.1, OpenSSL 1.0.1e-fips 11 Feb 2013
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: Connecting to ...4.63 [...4.63] port 2377.
debug1: Connection established.
```
Master lists itself and the first worker, but not the second:
```
03:~> docker node ls
ID                           HOSTNAME  MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
5yjryq2sup5qbpmptegbg78o6 *  ...03    Accepted    Ready   Active        Leader
c2vahnhcw2k2ik1z65v6zozyb    ...04    Accepted    Ready   Active        
```
The three nodes have the same OS, Kernel and Docker versions. They all have `firewalld` stopped and disabled. On all three nodes I get this warning when running `docker info`:
```
WARNING: bridge-nf-call-ip6tables is disabled
```
On only the second slave, the one which cannot join, I get an extra warning:
```
WARNING: bridge-nf-call-iptables is disabled
```
Notice `ip6tables` vs. `iptables`.

I tried leaving the swarm and joining it on the second worker but it made no difference.

**Comment by aaronlehmann** [comment]
Currently you'll need to pass `--listen-addr` to `swarm join` as well. This will no longer be necessary with some upcoming PRs that are blocked on #24237.

**Comment by rvernica** [comment]
Thanks for that, I added `--listen-addr`. The behavior is the same. The first worker joins successfully, the second worker fails to join:
```
05:~> docker swarm join --listen-addr ...4.65:2377 --secret ... \
>         --ca-hash ... \
>         ....4.63:2377
Error response from daemon: Timeout was reached before node was joined. Attempt to join the cluster will continue in the background. Use "docker info" command to see the current Swarm status of your node.
```
There is `rpc error: code = 1 desc = context canceled` reported on the second worker:
```
05:~> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.0-rc4
Storage Driver: devicemapper
 Pool Name: centos-dockerpool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 20.25 MB
 Data Space Total: 42.91 GB
 Data Space Available: 42.89 GB
 Metadata Space Used: 217.1 kB
 Metadata Space Total: 41.94 MB
 Metadata Space Available: 41.73 MB
 Thin Pool Minimum Free Space: 4.291 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge host null
Swarm: pending
 NodeID: 
 Error: rpc error: code = 1 desc = context canceled
 IsManager: No
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 5.65 GiB
Name: blu-05
ID: EM7E:PHL2:VOHH:TCP5:GID6:QKTH:WPHH:NDDV:HH5H:6LQ2:I6TC:ZU4C
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: ...
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Comment by nathanleclaire** [comment]
@rvernica Can you post Docker logs from the manager node and the failed join node around the time of attempted join?

**Comment by rvernica** [comment]
I figured it out. It was an user error. The clocks were not synchronized on the three nodes. The master and the worker which was able to join had very similar time. The worker which was not able to join was seven hours behind. I synchronized their clocks and now the two slaves can join successfully.

**Comment by thaJeztah** [comment]
^^ @sfsmithcha FYI

**Comment by nathanleclaire** [comment]
Ouch, clock skew.  Glad to hear you solved your issue @rvernica

**Comment by aaronlehmann** [comment]
I'm curious about what caused this, since I wasn't aware we depended on clock synchronization. I'm going to reopen it until we figure out what the issue is.

**Comment by aaronlehmann** [comment]
@diogomonica pointed out to me that TLS certificate expiration times would be affected by clock skew. But I don't think 7 hours of clock skew would be enough to trigger this, because the certificates are valid for a few months by default.

**Comment by aaronlehmann** [comment]
Actually, it could be TLS certificates. If the TLS certificate was issued after a machine's view of the current time, it won't be accepted. If this is what's happening, it looks like we're swallowing the TLS error and replacing it with a "deadline exceeded" error. We should make sure that this error is not swallowed.

**Comment by aaronlehmann** [comment]
I'm not sure this is related to TLS certificates after all. I tried making the manager issue certificates that are not valid yet, and I got a useful error immediately:

```
docker swarm join 192.168.200.246:4242
Error response from daemon: x509: certificate has expired or is not yet valid
```

If I get a chance, I'll try to spin up a few VMs and reproduce this with actual clock skew.

**Comment by aaronlehmann** [comment]
I was able to reproduce this on VMs with skewed clocks. It happens when a node joins a manager which has a certificate created in the "future".

The swallowing of the error is sort of a side effect of how GRPC's asynchronous dial works. It will keep retrying the connection, and the RPC call will block until it succeeds. We could (and should) put a time limit on the RPC using `context.WithDeadline`, but I'm not sure how to surface the error, which seems to go to the logs instead of the caller:

```
INFO[0001] grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: x509: certificate has expired or is not yet valid"; Reconnecting to "54.164.116.211:5000"
INFO[0002] grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: x509: certificate has expired or is not yet valid"; Reconnecting to "54.164.116.211:5000"
INFO[0004] grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: x509: certificate has expired or is not yet valid"; Reconnecting to "54.164.116.211:5000"
INFO[0008] grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: x509: certificate has expired or is not yet valid"; Reconnecting to "54.164.116.211:5000"
```

Does anyone know how to make grpc surface the error? It looks like switching to blocking `Dial` might do the trick (see code snip below). Are there any gotchas to be aware of with blocking `Dial`?

```go
        if c.dopts.block {
                if err := c.resetTransport(false); err != nil {
                        c.Close()
                        return nil, err
                }
                // Start to monitor the error status of transport.
                go c.transportMonitor()
        }
```

cc @LK4D4 @stevvooe @tonistiigi

**Comment by aaronlehmann** [comment]
Note that you'll see the same behavior if you try to join the wrong address. The connection will fail, but GRPC retries and swallows the error.

**Comment by aaronlehmann** [comment]
It looks like GRPC will swallow the error even with a blocking `Dial`. This code from `resetTransport` prints the error to the log and then `continue`s instead of returning the error. So I don't see how to surface the error.

```go
                        grpclog.Printf("grpc: Conn.resetTransport failed to create client transport: %v; Reconnecting to %q", err, cc.target)
                        continue
```

**Comment by nathanleclaire** [comment]
Do you all `SetLogger` with the GRPC log?  (https://godoc.org/google.golang.org/grpc/grpclog#SetLogger)   You could probably set it to a custom struct which embeds the default GRPC logger but has small wrapper methods to surface errors like this to the Docker log (or CLI).

**Comment by aaronlehmann** [comment]
@nathanleclaire: I'm not sure that will work because it has global scope and I want to surface the error from a specific connection attempt and return it to a specific Docker client that requested the connection.

**Comment by nathanleclaire** [comment]
@aaronlehmann  I know it may seem a bit weird, but you could iterate over the passed `...interface{}` to `Printf` and store it if it's an `error`, then ask your custom logger what the most recent error was (if on this branch of code and such an error seems suspect) to surface it.  Other alternative seems to be to submit a contribution to GRPC to surface this error somehow, but since it's nested in that internal method that might be tricky.

**Comment by aaronlehmann** [comment]
> I know it may seem a bit weird, but you could iterate over the passed ...interface{} to Printf and store it if it's an error, then ask your custom logger what the most recent error was (if on this branch of code and such an error seems suspect) to surface it.

This thought occurred to me, but this seems too hacky and racy. I wouldn't want to do something like this unless I could get some assurance that I'm not returning the wrong error.

> Other alternative seems to be to submit a contribution to GRPC to surface this error somehow, but since it's nested in that internal method that might be tricky.

So far this seems like the most reasonable approach, but it would take some time.

**Comment by nathanleclaire** [comment]
Yeah I hear you. Tough situation.  I wonder if some sort of "timestamp handshake" might be worth the trouble (there's an initial `Dial` and handshake that's not encrypted, right?  you could potentially do a basic sanity check then).

**Comment by aaronlehmann** [comment]
Relevant grpc-go issue is https://github.com/grpc/grpc-go/issues/622.

**Comment by stevvooe** [comment]
@aaronlehmann Are we using `WithBlock` anywhere? We should avoid it. `Dial` doesn't really "dial".

**Comment by aaronlehmann** [comment]
@stevvooe: We are not using `WithBlock` anywhere. I am proposing using it to work around grpc's `Dial` silently swallowing TLS errors. Unfortunately, that won't work until https://github.com/grpc/grpc-go/pull/768 is merged.

**Comment by aaronlehmann** [comment]
I'm moving this to 1.13.0 because 1.12.0 is already out, and it doesn't seem realistic to vendor a future GRPC release in a patch release, with all the changes that entails (such as replacing picker with balancer).

**Comment by aaronlehmann** [comment]
There is work underway on upgrading to a newer version of grpc-go in swarmkit: https://github.com/docker/swarmkit/pull/1460

Once this goes through, it should be possible to fix this.

**Comment by aluzzardi** [comment]
@LK4D4 - assigning to you since you're working on upgrading gRPC

**Comment by thaJeztah** [comment]
Linking to https://github.com/docker/docker/pull/26538, which is (I think) the PR where that is happening

**Comment by aluzzardi** [comment]
ping @LK4D4 (#26538 got merged)

**Comment by aaronlehmann** [comment]
Confirmed that this is fixed by #26538 🎉.

First replicated the problem with 1.12.1. Then upgraded the node which is trying to join to 1.13-dev. With the newer version, it shows this error immediately:

`Error response from daemon: rpc error: code = 13 desc = connection error: desc = "transport: x509: certificate has expired or is not yet valid"`

**Comment by svscorp** [comment]
@aaronlehmann @LK4D4 

I am having this issue on docker 17.10-ce. In my case it's no error code, message looks like this

````
Error response from daemon: rpc error: code = Unavailable desc = grpc: the connection is unavailable
````

RHEL 7.3, kernel 4.13.3-1.el7

**Comment by svscorp** [comment]
Ok, I figured out - it's because wildcard-whitelisted non-proxy hosts in the daemon http-proxy.conf configuration didn't work. Placed the exact node IPs and it goes through.
